File: java/src/dec_3/boj/연구소3_박재환.java
package dec_3.boj;

import java.util.*;
import java.io.*;

public class 연구소3_박재환 {
    static BufferedReader br;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        init();
        br.close();
    }

    /**
     * 바이러스는 활성 상태와 비활성 상태가 있다.
     * 초기에는 모두 비활성 상태이다.
     *
     * 활성 상태의 바이러스는 상하좌우로 인접한 모든 빈 칸으로 동시에 복제된다.-> 1초가 걸린다.
     *
     * 승원이는 바이러스 M개를 활성 상태로 변경하려한다.
     *
     * N x N
     * - 빈 칸( 0 )
     * - 벽 ( 1 )
     * - 바이러스 ( 2 )
     * 로 이루어져 있다.
     *
     * => M 개의 바이러스를 선택해서, 최단 시간 내에 바이러스를 모두 퍼뜨린다.
     */
    static StringTokenizer st;
    static int n, m;
    static int[][] board;
    static List<int[]> viruses;
    static int emptySpace;
    static void init() throws IOException {
        viruses = new ArrayList<>();
        emptySpace = 0;
        st = new StringTokenizer(br.readLine().trim());
        n = Integer.parseInt(st.nextToken());
        board = new int[n][n];
        m = Integer.parseInt(st.nextToken());

        for(int x=0; x<n; x++) {
            st = new StringTokenizer(br.readLine().trim());
            for(int y=0; y<n; y++) {
                board[x][y] = Integer.parseInt(st.nextToken());
                if(board[x][y] == 2) viruses.add(new int[] {x, y});
                else if(board[x][y] == 0) emptySpace++;
            }
        }
        // 바로 끝낼 수 있는 경우
        if(emptySpace == 0) {
            System.out.println(0);
            return;
        }

        solution();
    }

    static int minTime;
    static void solution() {
        minTime = Integer.MAX_VALUE;
        getVirusCombination(0,0, new int[m]);
        // 채울 수 없는 경우는 -1
        System.out.println(minTime == Integer.MAX_VALUE ? -1 : minTime);
    }

    static void getVirusCombination(int idx, int selectedIdx, int[] virusCandidate) {
        if(selectedIdx == m) {  // 조합 완성
            // 해당 조합으로, 바이러스를 퍼뜨릴 수 있는 최단 시간 구하기
            getMinTime(virusCandidate);
            return;
        }

        if(idx >= viruses.size()) return;   // 더 이상 조합을 만들 수 없음

        virusCandidate[selectedIdx] = idx;
        getVirusCombination(idx+1, selectedIdx+1, virusCandidate);
        getVirusCombination(idx+1, selectedIdx, virusCandidate);
    }

    static int[] dx = {0,1,0,-1};
    static int[] dy = {1,0,-1,0};
    static void getMinTime(int[] virusCandidate){
        Queue<int[]> q = new LinkedList<>();
        Queue<int[]> temp = new LinkedList<>();
        boolean[][] visited = new boolean[n][n];
        for(int idx : virusCandidate) {
            int[] point = viruses.get(idx);
            visited[point[0]][point[1]] = true;
            q.offer(point);
        }
        int[][] copyArr = copyArr();
        int time = 0;
        int spreadAreaCnt = 0;
        // 초 단위로 돌려야함
        // 큐를 모두 비우는 시점이 1초가 됨
        while(!q.isEmpty()) {
            // 가지치기
            if(time >= minTime) return;

            while(!q.isEmpty()) {
                int[] cur = q.poll();
                int curX = cur[0];
                int curY = cur[1];

                for(int dir=0; dir<4; dir++) {
                    int nx = curX + dx[dir];
                    int ny = curY + dy[dir];

                    if(nx < 0 || ny < 0 || nx >= n || ny >= n) continue;
                    if(copyArr[nx][ny] == 1 || visited[nx][ny]) continue;

                    if(copyArr[nx][ny] == 0) {  // 빈 칸으로 정상적으로 퍼지는 경우
                        spreadAreaCnt++;

                        if(spreadAreaCnt == emptySpace) {
                            minTime = Math.min(minTime, time + 1);
                            return;
                        }
                    }

                    copyArr[nx][ny] = 2;
                    visited[nx][ny] = true;
                    temp.offer(new int[] {nx, ny});
                }
            }

            if(temp.isEmpty()) break;

            while(!temp.isEmpty()) q.offer(temp.poll());
            time++;
        }
    }

    static int[][] copyArr() {
        int[][] copyArr = new int[n][n];
        for(int x=0; x<n; x++) {
            for(int y=0; y<n; y++) {
                copyArr[x][y] = board[x][y];
            }
        }
        return copyArr;
    }
}

File: python/src/dec_3/boj/연구소3_박재환.py
from collections import deque

MAX_INF = float('inf')
n, m = map(int, input().split(" "))
empty_space = 0
viruses = []
board = []
for x in range(n):
    y_arr = list(map(int, input().split(" ")))
    board.append(y_arr)
    for y in range(n):
        if board[x][y] == 0:
            empty_space+=1
        elif(board[x][y] == 2):
            viruses.append((x, y))


def solution(n, m, board, viruses, empty_space):
    min_time = MAX_INF
    # 1. 선택 가능한 바이러스 조합 찾기
    min_time = min(min_time, find_virus_combinations(0, 0, [0]*m, viruses, m, empty_space, n, board))
    print(-1 if min_time == MAX_INF else min_time)


def find_virus_combinations(idx, selected_idx, virus_combination, viruses, m, empty_space, n, board):
    if(selected_idx == m):
        # 바이러스 조합 완성
        return get_min_spread_time(virus_combination, board, n, empty_space)
    
    if(idx >= len(viruses) or (selected_idx + len(viruses)-idx) < m ):
        # 조합을 만들지 못하는 경우
        return MAX_INF

    virus_combination[selected_idx] = viruses[idx]
    t1 = find_virus_combinations(idx+1, selected_idx+1, virus_combination, viruses, m, empty_space, n, board)
    t2= find_virus_combinations(idx+1, selected_idx, virus_combination, viruses, m, empty_space, n, board)

    return min(t1, t2)

def get_min_spread_time(virus_combination, board, n, empty_space):
    q = deque()
    time_board = [list([-1]*n) for _ in range(n)]     # 배열 초기화
    for virus in virus_combination:
        x, y = virus
        q.append((x, y))
        time_board[x][y] = 0

    dx = [0,1,0,-1]
    dy = [1,0,-1,0]
    spread_cnt = 0
    max_time = -1

    while(q):
        x, y = q.popleft()
        cur_time = time_board[x][y]

        for dir in range(4):
            nx = x + dx[dir]
            ny = y + dy[dir]

            if nx < 0 or ny < 0 or nx >= n or ny >= n:
                continue
            if time_board[nx][ny] != -1 or board[nx][ny] == 1:
                continue

            time_board[nx][ny] = cur_time + 1
            q.append((nx, ny))

            # 빈 칸에 전이되는 경우 
            if board[nx][ny] == 0:
                spread_cnt += 1
                max_time = max(max_time, cur_time + 1)

                if spread_cnt == empty_space:
                    return max_time
    
    return MAX_INF

if __name__ == "__main__":
    if empty_space == 0:
        print(0)
    else:
        solution(n, m, board, viruses, empty_space)